compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -167,7 +166,7 @@ class ExpressionCodegen(
                 irFunction.markLineNumber(startOffset = irFunction is IrConstructor && irFunction.isPrimary)
             }
             val returnType = typeMapper.mapReturnType(irFunction)
-            result.coerce(returnType).materialize()
+            result.coerce(returnType, if (irFunction !is IrConstructor) irFunction.returnType else null).materialize()
```

Comment:
In this particular case, yes. Since this was confusing for a while now I've split the implementation of voidValue into two cases (a non-materialized value of type unit, and a promised default value).

This also made it possible to make the IrType field on PromisedValue non-nullable, but on the flip-side we now distinguish between materialized values on the stack and immaterial values.